File: tools/viz_check/cc/kernel_check.mjs
/* CC kernel cross-check, JS replay leg.
 *
 * Reads stdout JSON trace from /tmp/cc_kernel_check (the instrumented
 * C++ leg), runs the JS CC kernel on the same input, asserts that the
 * JS final-cluster decomposition is byte-equal to the C++ one.
 *
 * CC is fully deterministic (no RNG); replay is therefore trivial:
 * both legs run the same algorithm and must produce identical
 * components in identical iteration order.
 *
 * Run:
 *   /tmp/cc_kernel_check edge.csv com.csv js_out.csv > cpp.json 2>cpp.trace
 *   node tools/viz_check/cc/kernel_check.mjs cpp.json edge.csv com.csv [cpp.trace]
 *
 * When the optional cpp.trace path is supplied the replay also asserts
 * that the JS [TRACE-CC] probe stream (gated by globalThis.CC_DUMP_PROBES)
 * is byte-equal to cpp's. The CC P0 probe set covers:
 *   RIE_EDGE, RIE_SUM, CCW_ROOT, CCW_NEIGHS, CCW_VISIT, CCW_MEMBER,
 *   CCW_COMP_DONE, BUCKET, COMP_FLUSH, WCQ_POP, WriteClusterQueue.
 *
 * Exit code: 0 on byte-equal; 1 on any mismatch.
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
const __dirname = path.dirname(fileURLToPath(import.meta.url));

const args = process.argv.slice(2);
if (args.length < 3) {
  console.error("usage: kernel_check.mjs <cpp_trace.json> <edge.csv> <com.csv> [cpp.trace]");
  process.exit(2);
}
const [cppTracePath, edgePath, comPath, cppTraceLogPath] = args;

const cpp = JSON.parse(fs.readFileSync(cppTracePath, "utf8"));

// Load JS kernel against the same input. Enable per-step [TRACE-CC]
// probes BEFORE importing cc.js so the module-level _ccProbesEnabled()
// gate sees the flag at first call.
globalThis.window = globalThis;
globalThis.window.COMDET = { FIXTURE: { nodes: [], edges: [], gt: [] } };
globalThis.CC_DUMP_PROBES = true;

// Capture JS probe emissions (stderr) and assert byte-equality vs cpp.
const jsTraceLines = [];
const origStderrWrite = process.stderr.write.bind(process.stderr);
const origConsoleError = console.error;
console.error = function (msg) {
  if (typeof msg === "string" && msg.startsWith("[TRACE-CC]")) {
    jsTraceLines.push(msg);
    return;
  }
  origConsoleError.apply(console, arguments);
};

const WEB = path.join(__dirname, "../../../../../web/vltanh.github.io/comdet/js");
await import(path.join(WEB, "mincut.js"));
await import(path.join(WEB, "louvain/louvain.js"));
await import(path.join(WEB, "leiden/leiden.js"));
await import(path.join(WEB, "comdet/page_helpers.js"));
await import(path.join(WEB, "cc/cc.js"));

function loadGraph(edgePath, comPath) {
  const edgeText = fs.readFileSync(edgePath, "utf8").trim().split(/\r?\n/);
  const com = fs.readFileSync(comPath, "utf8").trim().split(/\r?\n/);
  const orig_to_new = new Map();
  const edges = [];
  // First line is header.
  for (let i = 1; i < edgeText.length; i++) {
    if (!edgeText[i]) continue;
    const cols = edgeText[i].split(/[,\t ]/);
    const s = cols[0], t = cols[1];
    if (!orig_to_new.has(s)) orig_to_new.set(s, orig_to_new.size);
    if (!orig_to_new.has(t)) orig_to_new.set(t, orig_to_new.size);
    edges.push([orig_to_new.get(s), orig_to_new.get(t)]);
  }
  const partition = new Int32Array(orig_to_new.size);
  for (let i = 0; i < partition.length; i++) partition[i] = -2147483648;
  for (let i = 1; i < com.length; i++) {
    if (!com[i]) continue;
    const cols = com[i].split(/[,\t ]/);
    const newId = orig_to_new.get(cols[0]);
    if (newId == null) continue;
    partition[newId] = parseInt(cols[1], 10);
  }
  return {
    nodes: Array.from({ length: orig_to_new.size }, (_, i) => i),
    edges,
    membership: partition,
  };
}

const G = loadGraph(edgePath, comPath);
const fixture = { nodes: G.nodes, edges: G.edges, gt: Array.from(G.membership) };

const r = COMDET.CC.runCC(G.membership, { fixture });

// Compare component-by-component, byte-equal.
// Canonical's GetConnectedComponents filters singletons (`csize > 1`) at
// constrained.h:409 — comparison must mirror that filter on the JS side.
const cppComps = cpp.components;
const jsComps = r.allComps.filter(function (c) { return c.length > 1; });
let pass = true;
const report = [];
report.push(`cpp components: ${cppComps.length}`);
report.push(`js  components: ${jsComps.length}`);
if (cppComps.length !== jsComps.length) pass = false;
const N = Math.max(cppComps.length, jsComps.length);
for (let i = 0; i < N; i++) {
  const c = cppComps[i] || [];
  const j = jsComps[i] || [];
  const cStr = JSON.stringify(c);
  const jStr = JSON.stringify(j);
  if (cStr !== jStr) {
    pass = false;
    report.push(`  comp[${i}] DIFFER cpp=${cStr.slice(0, 60)} js=${jStr.slice(0, 60)}`);
  } else {
    report.push(`  comp[${i}] match size=${c.length}`);
  }
}

// Optional: per-step trajectory diff vs cpp's [TRACE-CC] stream.
if (cppTraceLogPath && fs.existsSync(cppTraceLogPath)) {
  const cppTrace = fs.readFileSync(cppTraceLogPath, "utf8")
                     .split(/\r?\n/)
                     .filter((l) => l.startsWith("[TRACE-CC]"));
  // Drop preamble lines that aren't per-step (PIPELINE_START / loaded /
  // ReadCommunities / GetConnectedComponents summary / DONE_QUEUE /
  // PIPELINE_END / per-comp summary rows). These reference cpp-only
  // bookkeeping (filenames, build mode, queue sizes) the JS port doesn't
  // emit. The per-step rows (RIE_EDGE, CCW_*, CCW_MEMBER, BUCKET,
  // COMP_FLUSH, WCQ_POP, RIE_SUM, WriteClusterQueue) are required to
  // match byte-for-byte.
  const STEP_TAGS = ["RIE_EDGE", "RIE_SUM", "CCW_ROOT", "CCW_NEIGHS",
                     "CCW_VISIT", "CCW_MEMBER", "CCW_COMP_DONE",
                     "BUCKET", "COMP_FLUSH", "WCQ_POP",
                     "WriteClusterQueue"];
  function isStep(line) {
    for (const tag of STEP_TAGS) {
      if (line.startsWith(`[TRACE-CC] ${tag}`)) return true;
    }
    return false;
  }
  const cppStep = cppTrace.filter(isStep);
  const jsStep = jsTraceLines.filter(isStep);
  report.push(`cpp trajectory steps: ${cppStep.length}`);
  report.push(`js  trajectory steps: ${jsStep.length}`);
  if (cppStep.length !== jsStep.length) {
    pass = false;
    report.push(`TRAJECTORY COUNT MISMATCH`);
  }
  const M = Math.min(cppStep.length, jsStep.length);
  let firstDiff = -1;
  for (let i = 0; i < M; i++) {
    if (cppStep[i] !== jsStep[i]) {
      pass = false;
      if (firstDiff < 0) {
        firstDiff = i;
        report.push(`  trajectory[${i}] DIFFER`);
        report.push(`    cpp=${cppStep[i]}`);
        report.push(`    js =${jsStep[i]}`);
      }
    }
  }
  if (firstDiff < 0 && cppStep.length === jsStep.length) {
    report.push(`trajectory: ${cppStep.length} steps byte-equal`);
  }
}

console.log = origConsoleError; // restore raw stdout for the summary
console.error = origConsoleError;
process.stdout.write(report.join("\n") + "\n");
if (!pass) {
  process.stderr.write("FAIL: CC tracer vs JS replay produced different output.\n");
  process.exit(1);
}
process.stdout.write("PASS: CC tracer == JS replay (byte-equal components + trajectory).\n");

File: tools/viz_check/cc/kernel_check.py
"""CC kernel cross-check driver.

Three legs per fixture:

1. **Canonical leg** — `constrained_clustering MincutOnly --connectedness-criterion 0`.
2. **Instrumented C++ leg** — /tmp/cc_kernel_check; verbatim copy of the CC pipeline
   (mincut_only.cpp Simple branch + GetConnectedComponents) with stderr trace at
   every key state. Asserts byte-equal CSV output vs canonical.
3. **JS replay leg** — tools/viz_check/cc/kernel_check.mjs; consumes the C++ tracer's
   stdout JSON, runs comdet/js/cc/cc.js on the same input, asserts byte-equal
   per-cluster node lists.

CC has no RNG. The three legs must produce identical output.

Run:
    python tools/viz_check/cc/kernel_check.py
    python tools/viz_check/cc/kernel_check.py --verbose
"""
from __future__ import annotations

import argparse
import json
import shutil
import subprocess
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
REPO = HERE.parent.parent.parent  # community-detection/
BIN = REPO / "constrained-clustering" / "build" / "bin" / "constrained_clustering"
TRACER = Path("/tmp/cc_kernel_check")
JS_REPLAY = HERE / "kernel_check.mjs"


def build_tracer():
    rc = subprocess.run(["bash", str(HERE / "instrumented" / "build.sh")],
                        capture_output=True, text=True)
    if rc.returncode != 0:
        sys.stderr.write(rc.stdout + rc.stderr)
        sys.exit(2)


def emit_fixture_inputs(out_dir: Path):
    """Emit edge.csv + com_gt.csv from the comdet 32-node fixture."""
    fixture_emitter = REPO / "tests" / "cd_verify" / "emit_fixture.js"
    rc = subprocess.run(["node", str(fixture_emitter), str(out_dir)],
                        capture_output=True, text=True)
    if rc.returncode != 0:
        sys.stderr.write(rc.stdout + rc.stderr)
        sys.exit(2)


def run_canonical(edge: Path, com: Path, out: Path) -> str:
    log = out.with_suffix(".log")
    rc = subprocess.run([str(BIN), "MincutOnly",
                         "--edgelist", str(edge),
                         "--existing-clustering", str(com),
                         "--output-file", str(out),
                         "--log-file", str(log),
                         "--connectedness-criterion", "0",
                         "--num-processors", "1"],
                        capture_output=True, text=True)
    if rc.returncode != 0:
        raise RuntimeError(f"binary failed: {rc.stderr}")
    return out.read_text()


def run_tracer(edge: Path, com: Path, out: Path) -> tuple[str, str, str]:
    rc = subprocess.run([str(TRACER), str(edge), str(com), str(out)],
                        capture_output=True, text=True)
    if rc.returncode != 0:
        raise RuntimeError(f"tracer failed: {rc.stderr}")
    return out.read_text(), rc.stdout, rc.stderr


def run_js_replay(trace_json: Path, edge: Path, com: Path) -> tuple[bool, str]:
    rc = subprocess.run(["node", str(JS_REPLAY), str(trace_json),
                         str(edge), str(com)],
                        capture_output=True, text=True)
    return rc.returncode == 0, rc.stdout + rc.stderr


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--verbose", action="store_true")
    args = ap.parse_args()

    if not BIN.exists():
        sys.exit(f"missing binary: {BIN}; build first")

    build_tracer()

    fixtures = []
    # 1. 32-node comdet fixture.
    work = REPO / "tests" / "cd_verify"
    work.mkdir(parents=True, exist_ok=True)
    emit_fixture_inputs(work)
    fixtures.append(("fixture32", work / "edge.csv", work / "com_gt.csv"))
    # 2. dnc with leiden-mod input clustering (already staged in cd_verify/).
    if (work / "dnc_edge.csv").exists() and (work / "dnc_com.csv").exists():
        fixtures.append(("dnc", work / "dnc_edge.csv", work / "dnc_com.csv"))

    failures = 0
    for name, edge, com in fixtures:
        print(f"\n=== {name} ===")
        canon_out = work / f"{name}_canon.csv"
        tracer_out = work / f"{name}_tracer.csv"
        tracer_json = work / f"{name}_tracer.json"
        canon_csv = run_canonical(edge, com, canon_out)
        tracer_csv, tracer_stdout, tracer_stderr = run_tracer(edge, com, tracer_out)
        tracer_json.write_text(tracer_stdout)
        if args.verbose:
            print("[tracer trace]")
            print(tracer_stderr)

        # Leg 1 vs Leg 2: byte-equal CSV.
        if canon_csv == tracer_csv:
            print(f"  canonical == tracer: PASS (byte-equal CSV, {len(canon_csv.splitlines())} lines)")
        else:
            print(f"  canonical != tracer: FAIL")
            failures += 1

        # Leg 2 vs Leg 3: JS replay.
        ok, replay_log = run_js_replay(tracer_json, edge, com)
        last = replay_log.strip().splitlines()[-1] if replay_log.strip() else "(no output)"
        if ok:
            print(f"  tracer == js_replay: PASS ({last})")
        else:
            print(f"  tracer != js_replay: FAIL")
            print(replay_log)
            failures += 1

    print()
    if failures:
        print(f"OVERALL: FAIL ({failures} failures)")
        sys.exit(1)
    print("OVERALL: PASS")


if __name__ == "__main__":
    main()
